File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/pure/RunStatus.kt
// ── spine/pure/run-status — SESSION-GLOBAL, boundary-owned ─────────────────
// Idle/Working describe the turn; Degraded/Error describe the SESSION: a budget
// exceeded, an append that failed, a turn that threw.
//
// Degraded and Error may be constructed only in spine/boundary/** (gate check C6).
// A block cannot reach the session banner at all — which is the structural fix for
// 12.4, not a convention an author has to keep.

package adr.spine.pure

sealed class RunStatus {
    data object Idle : RunStatus()

    data class Working(val step: Int) : RunStatus()

    data class Degraded(val cause: String) : RunStatus()

    data class Error(val fault: String) : RunStatus()
}

File: agent-driven-architecture-main/examples/typescript/src/spine/pure/run-status.ts
// ── spine/pure/run-status — SESSION-GLOBAL, boundary-only (12.4) ────────────
// A closed set describing the SESSION, not an item. `Degraded` and `Error` may
// be constructed only inside `spine/boundary/**` — a budget exceeded, an append
// that failed, a turn that threw. Gate check C6 denies both constructors under
// `blocks/**`, so the 12.4 bug (a per-item rejection hijacking the session
// banner) cannot be written again.

export type RunStatusKind = "Idle" | "Working" | "Degraded" | "Error";

export interface RunStatusBase {
  readonly kind: RunStatusKind;
}

export interface Idle extends RunStatusBase {
  readonly kind: "Idle";
}
export interface Working extends RunStatusBase {
  readonly kind: "Working";
  readonly step: number;
}
export interface Degraded extends RunStatusBase {
  readonly kind: "Degraded";
  readonly cause: string;
}
export interface Errored extends RunStatusBase {
  readonly kind: "Error";
  readonly fault: string;
}

export type RunStatus = Idle | Working | Degraded | Errored;

export const idle: Idle = { kind: "Idle" };

export function working(step: number): Working {
  return { kind: "Working", step };
}

/** boundary-only (check C6) */
export function degraded(cause: string): Degraded {
  return { kind: "Degraded", cause };
}

/** boundary-only (check C6) */
export function errored(fault: string): Errored {
  return { kind: "Error", fault };
}
